failure: MSB4223: A node of the required type InProc could not be created.

I have a test project that I am upgrading from MSBuildProjectCreation 3.0.5 to 6.2.4
This test project fails with "error MSB4223: A node of the required type InProc could not be created."

The scenario that this test tries to cover should be similar to a solution building multiple projects.
(in our case we are actually testing the building of 2 solutions in the right order based on dependencies defined in our custom MSBuild targets files)

I saw in the changelogs that there was a recent change with outproc build https://github.com/jeffkl/MSBuildProjectCreator/pull/111

This includes a hardcoded setting with `MaxNodeCount = 1`, I am wondering if that could be the explanation of our specific failure since we have on project file that is using the MSBuild task to build multiple other "projects"?

Now I get an error `SolutionOne.sln(2,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 2, position 1. [SolutionOne.sln]`

This solution is a hardcoded sln file in the testcase:

```sln
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio Version 16
VisualStudioVersion = 16.0.29102.190
MinimumVisualStudioVersion = 10.0.40219.1
Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ProjectOne", "ProjectOne\ProjectOne.csproj", "{9B189FD2-59BE-4327-97D8-3170AF70C02F}"
EndProject
Global
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|Win32 = Debug|Win32
		Release|Win32 = Release|Win32
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{9B189FD2-59BE-4327-97D8-3170AF70C02F}.Debug|Win32.ActiveCfg = Debug|x86
		{9B189FD2-59BE-4327-97D8-3170AF70C02F}.Debug|Win32.Build.0 = Debug|x86
		{9B189FD2-59BE-4327-97D8-3170AF70C02F}.Release|Win32.ActiveCfg = Release|x86
		{9B189FD2-59BE-4327-97D8-3170AF70C02F}.Release|Win32.Build.0 = Release|x86
	EndGlobalSection
	GlobalSection(SolutionProperties) = preSolution
		HideSolutionNode = FALSE
	EndGlobalSection
	GlobalSection(ExtensibilityGlobals) = postSolution
		SolutionGuid = {471DDE0F-A01E-4350-B423-6599E9FF7639}
	EndGlobalSection
EndGlobal
```

This worked fine previously (but please note that I am also upgrading from Visual Studio 2019 16.9 to 16.11).

Opening the solution from VisualStudio 2019 itself is ok, and running through MSBuild (commandline) also seems to work ok

--- a/README.md
+++ b/README.md
@@ -44,10 +44,13 @@ The resulting project would look like this:
 ## Building Projects
 Use the `TryBuild` methods to build your projects.  `TryBuild` returns a `BuildOutput` object which captures the build output for you.
 
-This example creates a project that logs a message, executes `TryBuild`, and asserts some conditions against the build output.
+**Note:** Projects are built in a different process to avoid assembly load conflicts so projects must be saved before being built.
+
+This example creates a project that logs a message, saves the project, executes `TryBuild`, and asserts some conditions against the build output.
 ```C#
 ProjectCreator.Create()
     .TaskMessage("Hello, World!")
+    .Save(path)
     .TryBuild(out bool success, out BuildOutput log);
 
 Assert.True(success);

--- a/src/Microsoft.Build.Utilities.ProjectCreation.UnitTests/BuildTests.cs
+++ b/src/Microsoft.Build.Utilities.ProjectCreation.UnitTests/BuildTests.cs
@@ -7,9 +7,11 @@ using Microsoft.Build.Execution;
 using Microsoft.Build.Framework;
 using Microsoft.Build.Logging;
 using Shouldly;
+using System;
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using System.Runtime.InteropServices;
 using Xunit;
 
 namespace Microsoft.Build.Utilities.ProjectCreation.UnitTests
@@ -21,7 +23,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation.UnitTests
         {
             ProjectCreator
                 .Create(
-                    Path.Combine(TestRootPath, "project1.proj"),
+                    Path.Combine(TestRootPath, "project2.proj"),
                     defaultTargets: "Build")
                 .Target("Build")
                 .Target("Restore")
@@ -76,6 +78,53 @@ namespace Microsoft.Build.Utilities.ProjectCreation.UnitTests
             resultWithGlobalProperties.ShouldBeTrue();
 
             buildOutputWithGlobalProperties.MessageEvents.High.ShouldHaveSingleItem(buildOutputWithGlobalProperties.GetConsoleLog()).Message.ShouldBe("Value = D7BBABDFB2D142D3A75E0C1A33E33780", buildOutputWithGlobalProperties.GetConsoleLog());
+        }
+
+        [Fact]
+        public void ProjectWithNoPathBuildThrowsInvalidOperationException()
+        {
+            InvalidOperationException exception = Should.Throw<InvalidOperationException>(() =>
+            {
+                ProjectCreator.Templates.LogsMessage("6E83EF78-959F-45A2-9FE3-08BAD99C0F92")
+                    .TryBuild(out bool _);
+            });
+
+            exception.Message.ShouldBe("Project has not been given a path to save to.");
+        }
+
+        [Fact]
+        public void CanBuildLotsOfProjects()
+        {
+            int maxBuilds = Environment.ProcessorCount * 2;
+
+            List<ProjectCreator> projects = new List<ProjectCreator>(maxBuilds);
+
+            for (int i = 0; i < maxBuilds; i++)
+            {
+                projects.Add(
+                    ProjectCreator.Create(GetTempProjectPath())
+                        .Target("Build")
+                        .Task(
+                            "Exec",
+                            parameters: new Dictionary<string, string>
+                            {
+                                ["Command"] = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "ping 127.0.0.1 -n 2 >NUL" : "sleep 2s",
+                            })
+                        .Save());
+            }
+
+            ProjectCreator.Create(path: GetTempProjectPath())
+                .ForEach(projects, (project, creator) => creator.ItemProjectReference(project))
+                .Target("Build")
+                .Task("MSBuild", parameters: new Dictionary<string, string>
+                {
+                    ["Projects"] = "@(ProjectReference)",
+                    ["BuildInParallel"] = bool.TrueString,
+                })
+                .Save()
+                .TryBuild(out bool result, out BuildOutput buildOutput);
+
+            result.ShouldBeTrue(buildOutput.GetConsoleLog());
         }
 
         [Fact]

--- a/src/Microsoft.Build.Utilities.ProjectCreation.UnitTests/PackageFeedTests/PackageFeedTests.cs
+++ b/src/Microsoft.Build.Utilities.ProjectCreation.UnitTests/PackageFeedTests/PackageFeedTests.cs
@@ -38,7 +38,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation.UnitTests.PackageFeedTests
         [Fact]
         public void Test1()
         {
-            var targetFrameworks = new[]
+            string[] targetFrameworks = new[]
             {
                 "net45",
                 "net46",

--- a/src/Microsoft.Build.Utilities.ProjectCreation.UnitTests/PackageRepositoryTests/RepositoryTests.cs
+++ b/src/Microsoft.Build.Utilities.ProjectCreation.UnitTests/PackageRepositoryTests/RepositoryTests.cs
@@ -27,9 +27,9 @@ namespace Microsoft.Build.Utilities.ProjectCreation.UnitTests.PackageRepositoryT
                     .Library("netstandard2.0"))
             {
                 ProjectCreator.Templates.SdkCsproj(
+                        path: Path.Combine(TestRootPath, "ClassLibraryA", "ClassLibraryA.csproj"),
                         targetFramework: TargetFramework)
                     .ItemPackageReference(packageA)
-                    .Save(Path.Combine(TestRootPath, "ClassLibraryA", "ClassLibraryA.csproj"))
                     .TryBuild(restore: true, out bool result, out BuildOutput buildOutput);
 
                 result.ShouldBeTrue(buildOutput.GetConsoleLog());

--- a/src/Microsoft.Build.Utilities.ProjectCreation.UnitTests/TestBase.cs
+++ b/src/Microsoft.Build.Utilities.ProjectCreation.UnitTests/TestBase.cs
@@ -54,5 +54,12 @@ namespace Microsoft.Build.Utilities.ProjectCreation.UnitTests
         {
             return Path.Combine(TestRootPath, $"{Path.GetRandomFileName()}{extension ?? string.Empty}");
         }
+
+        protected string GetTempProjectPath(string extension = null)
+        {
+            DirectoryInfo tempDirectoryInfo = Directory.CreateDirectory(Path.Combine(TestRootPath, Path.GetRandomFileName()));
+
+            return Path.Combine(tempDirectoryInfo.FullName, $"{Path.GetRandomFileName()}{extension ?? string.Empty}");
+        }
     }
 }

--- a/src/Microsoft.Build.Utilities.ProjectCreation/BuildManagerHost.cs
+++ b/src/Microsoft.Build.Utilities.ProjectCreation/BuildManagerHost.cs
@@ -30,6 +30,11 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         /// </summary>
         private static readonly Lazy<FileInfo> MSBuildAssemblyFullPathLazy = new Lazy<FileInfo>(() => new FileInfo(typeof(BuildManager).Assembly.Location));
 
+        static BuildManagerHost()
+        {
+            Environment.SetEnvironmentVariable("MSBUILDNOINPROCNODE", "1");
+        }
+
         /// <summary>
         /// Executes a build for the specified project.
         /// </summary>
@@ -43,7 +48,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         {
             BuildRequestData buildRequestData = new BuildRequestData(
                 projectFullPath,
-                globalProperties,
+                globalProperties ?? new Dictionary<string, string>(),
                 toolsVersion: null,
                 targets ?? Array.Empty<string>(),
                 hostServices: null,
@@ -144,9 +149,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         {
             lock (LockObject)
             {
-                using BuildManager buildManager = new BuildManager();
-
-                MuxLogger muxLogger = new MuxLogger()
+                MuxLogger muxLogger = new MuxLogger
                 {
                     Verbosity = LoggerVerbosity.Diagnostic,
                 };
@@ -155,7 +158,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
                 {
                     DisableInProcNode = true,
                     EnableNodeReuse = false,
-                    MaxNodeCount = 1,
+                    MaxNodeCount = Environment.ProcessorCount,
                     ResetCaches = true,
                 };
 
@@ -175,11 +178,11 @@ namespace Microsoft.Build.Utilities.ProjectCreation
                     forwardingLoggerRecord,
                 };
 
-                buildManager.BeginBuild(buildParameters);
+                BuildManager.DefaultBuildManager.BeginBuild(buildParameters);
 
                 try
                 {
-                    BuildSubmission buildSubmission = buildManager.PendBuildRequest(buildRequestData);
+                    BuildSubmission buildSubmission = BuildManager.DefaultBuildManager.PendBuildRequest(buildRequestData);
 
                     foreach (ILogger logger in loggers)
                     {
@@ -188,7 +191,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
 
                     try
                     {
-                        SetCurrentHost(buildManager);
+                        SetCurrentHost(BuildManager.DefaultBuildManager);
 
                         BuildResult buildResult = buildSubmission.Execute();
 
@@ -206,7 +209,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
                 }
                 finally
                 {
-                    buildManager.EndBuild();
+                    BuildManager.DefaultBuildManager.EndBuild();
                 }
             }
         }

--- a/src/Microsoft.Build.Utilities.ProjectCreation/ExtensionMethods.cs
+++ b/src/Microsoft.Build.Utilities.ProjectCreation/ExtensionMethods.cs
@@ -5,6 +5,7 @@
 using NuGet.LibraryModel;
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
 
 namespace Microsoft.Build.Utilities.ProjectCreation
@@ -20,6 +21,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         /// <typeparam name="T">The type of the object.</typeparam>
         /// <param name="item">The item to make into an <see cref="IEnumerable{T}"/>.</param>
         /// <returns>The current object as an <see cref="IEnumerable{T}"/>.</returns>
+        [DebuggerStepThrough]
         public static IEnumerable<T> AsEnumerable<T>(this T item)
             where T : class
         {
@@ -40,6 +42,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         /// </summary>
         /// <param name="libraryIncludeFlags">The <see cref="LibraryIncludeFlags" /> to enumerate.</param>
         /// <returns>Nothing if <paramref name="libraryIncludeFlags" /> is <see cref="LibraryIncludeFlags.None" />, <see cref="LibraryIncludeFlags.All" />, or all flags.</returns>
+        [DebuggerStepThrough]
         public static IEnumerable<LibraryIncludeFlags> EnumerateExcludeFlags(this LibraryIncludeFlags libraryIncludeFlags)
         {
             if (libraryIncludeFlags == LibraryIncludeFlags.All)
@@ -63,6 +66,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         /// </summary>
         /// <param name="libraryIncludeFlags">The <see cref="LibraryIncludeFlags" /> to enumerate.</param>
         /// <returns>Nothing if <paramref name="libraryIncludeFlags" /> is <see cref="LibraryIncludeFlags.All" />, <see cref="LibraryIncludeFlags.None" />, or all flags.</returns>
+        [DebuggerStepThrough]
         public static IEnumerable<LibraryIncludeFlags> EnumerateIncludeFlags(this LibraryIncludeFlags libraryIncludeFlags)
         {
             if (libraryIncludeFlags == LibraryIncludeFlags.None)
@@ -87,6 +91,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         /// <param name="first">The first dictionary and all of its values to start with.</param>
         /// <param name="second">The second dictionary to merge with the first and override its values.</param>
         /// <returns>A merged <see cref="IDictionary{String,String}"/> with the values of the first dictionary overridden by the second.</returns>
+        [DebuggerStepThrough]
         public static IDictionary<string, string> Merge(this IDictionary<string, string> first, IDictionary<string, string> second)
         {
             return Merge(first, second, StringComparer.OrdinalIgnoreCase);
@@ -99,6 +104,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         /// <param name="second">The second dictionary to merge with the first and override its values.</param>
         /// <param name="comparer">The <see cref="IEqualityComparer{String}"/> implementation to use when comparing keys, or null to use the default <see cref="IEqualityComparer{String}"/> for the type of the key. </param>
         /// <returns>A merged <see cref="IDictionary{String,String}"/> with the values of the first dictionary overridden by the second.</returns>
+        [DebuggerStepThrough]
         public static IDictionary<string, string> Merge(this IDictionary<string, string> first, IDictionary<string, string> second, IEqualityComparer<string> comparer)
         {
             Dictionary<string, string> result = first == null
@@ -119,6 +125,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         /// <typeparam name="T">The type of the object.</typeparam>
         /// <param name="item">The item to make into an array.</param>
         /// <returns>An array of T objects.</returns>
+        [DebuggerStepThrough]
         public static T[] ToArrayWithSingleElement<T>(this T item)
             where T : class
         {

--- a/src/Microsoft.Build.Utilities.ProjectCreation/Package.cs
+++ b/src/Microsoft.Build.Utilities.ProjectCreation/Package.cs
@@ -205,8 +205,8 @@ namespace Microsoft.Build.Utilities.ProjectCreation
                 _dependencies[targetFramework] = packageDependencies;
             }
 
-            var includes = new List<string>();
-            var excludes = new List<string>();
+            List<string> includes = new List<string>();
+            List<string> excludes = new List<string>();
 
             LibraryIncludeFlags effectiveInclude = include & ~exclude & ~suppressParent;
 
@@ -299,7 +299,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
 
             using (Stream stream = File.Create(FullPath))
             {
-                foreach (var file in _files.Values)
+                foreach (IPackageFile file in _files.Values)
                 {
                     _packageBuilder.Files.Add(file);
                 }

--- a/src/Microsoft.Build.Utilities.ProjectCreation/ProjectCreator.Build.cs
+++ b/src/Microsoft.Build.Utilities.ProjectCreation/ProjectCreator.Build.cs
@@ -2,7 +2,6 @@
 //
 // Licensed under the MIT license.
 
-using Microsoft.Build.Evaluation;
 using Microsoft.Build.Execution;
 using System;
 using System.Collections.Generic;
@@ -316,7 +315,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         /// <returns>The current <see cref="ProjectCreator"/>.</returns>
         public ProjectCreator TryBuild(string[] targets, IDictionary<string, string> globalProperties, out bool result, out BuildOutput buildOutput, out IDictionary<string, TargetResult> targetOutputs)
         {
-            return TryBuild(restore: false, targets, out result, out buildOutput, out targetOutputs);
+            return TryBuild(restore: false, targets, globalProperties, out result, out buildOutput, out targetOutputs);
         }
 
         /// <summary>
@@ -490,9 +489,9 @@ namespace Microsoft.Build.Utilities.ProjectCreation
 
         private void Build(bool restore, string[] targets, IDictionary<string, string> globalProperties, BuildOutput buildOutput, out bool result, out IDictionary<string, TargetResult> targetOutputs)
         {
-            Save();
-
             targetOutputs = null;
+
+            Save();
 
             if (restore)
             {
@@ -504,20 +503,12 @@ namespace Microsoft.Build.Utilities.ProjectCreation
                 }
             }
 
-            ProjectInstance projectInstance;
-
-            if (globalProperties != null)
-            {
-                TryGetProject(out Project project, globalProperties);
-
-                projectInstance = project.CreateProjectInstance();
-            }
-            else
-            {
-                projectInstance = ProjectInstance;
-            }
-
-            BuildResult buildResult = BuildManagerHost.Build(projectInstance, targets ?? projectInstance.DefaultTargets.ToArray(), globalProperties, new List<Framework.ILogger>(ProjectCollection.Loggers.Concat(buildOutput.AsEnumerable())), BuildRequestDataFlags.ReplaceExistingProjectInstance);
+            BuildResult buildResult = BuildManagerHost.Build(
+                FullPath,
+                targets,
+                globalProperties,
+                new List<Framework.ILogger>(ProjectCollection.Loggers.Concat(buildOutput.AsEnumerable())),
+                BuildRequestDataFlags.None);
 
             result = buildResult.OverallResult == BuildResultCode.Success;
 
@@ -543,15 +534,19 @@ namespace Microsoft.Build.Utilities.ProjectCreation
             globalProperties["ExcludeRestorePackageImports"] = "true";
             globalProperties["MSBuildRestoreSessionId"] = Guid.NewGuid().ToString("D");
 
-            BuildResult buildResult = BuildManagerHost.Build(FullPath, new[] { "Restore" }, globalProperties, new List<Framework.ILogger>(ProjectCollection.Loggers.Concat(buildOutput.AsEnumerable())), BuildRequestDataFlags.ClearCachesAfterBuild | BuildRequestDataFlags.SkipNonexistentTargets | BuildRequestDataFlags.IgnoreMissingEmptyAndInvalidImports);
-
-            Project.MarkDirty();
-
-            Project.ReevaluateIfNecessary();
-
-            _projectInstance = Project.CreateProjectInstance();
+            BuildResult buildResult = BuildManagerHost.Build(
+                FullPath,
+                new[] { "Restore" },
+                globalProperties,
+                new List<Framework.ILogger>(ProjectCollection.Loggers.Concat(buildOutput.AsEnumerable())),
+                BuildRequestDataFlags.ClearCachesAfterBuild | BuildRequestDataFlags.SkipNonexistentTargets | BuildRequestDataFlags.IgnoreMissingEmptyAndInvalidImports);
 
             targetOutputs = buildResult.ResultsByTarget;
+
+            if (buildResult.Exception != null)
+            {
+                throw buildResult.Exception;
+            }
 
             result = buildResult.OverallResult == BuildResultCode.Success;
         }

--- a/src/Microsoft.Build.Utilities.ProjectCreation/ProjectCreator.Project.cs
+++ b/src/Microsoft.Build.Utilities.ProjectCreation/ProjectCreator.Project.cs
@@ -48,7 +48,7 @@ namespace Microsoft.Build.Utilities.ProjectCreation
         /// <summary>
         /// Gets the <see cref="ProjectInstance" /> for the current project.
         /// </summary>
-        public ProjectInstance ProjectInstance => _projectInstance ?? (_projectInstance = Project.CreateProjectInstance());
+        public ProjectInstance ProjectInstance => _projectInstance ??= Project.CreateProjectInstance();
 
         /// <summary>
         /// Gets a <see cref="Project"/> instance from the current project.

--- a/version.json
+++ b/version.json
@@ -1,5 +1,5 @@
 {
-  "version": "6.2",
+  "version": "6.3",
   "assemblyVersion": "1.0",
   "buildNumberOffset": -1,
   "nugetPackageVersion": {

